File: step_definitions/Verify system performance with maximum concurrent users accessing the catalog.java
package com.cucumber.steps;

import com.page_objects.PerformancePage;
import io.cucumber.java.en.Given;
import io.cucumber.java.en.Then;
import io.cucumber.java.en.When;

public class PerformanceSteps extends PerformancePage {

    @Given("the system is prepared for maximum concurrent user load")
    public void theSystemIsPreparedForMaximumConcurrentUserLoad() {
        prepareSystemForMaxLoad();
    }

    @When("concurrent user sessions are initiated to access the product catalog")
    public void concurrentUserSessionsAreInitiatedToAccessTheProductCatalog() {
        initiateConcurrentSessions();
    }

    @Then("the system handles multiple user sessions without crashing")
    public void theSystemHandlesMultipleUserSessionsWithoutCrashing() {
        verifySystemStability();
    }

    @Then("the product catalog must be accessible")
    public void theProductCatalogMustBeAccessible() {
        verifyCatalogAccessibility();
    }

    @Given("the system is under peak load conditions")
    public void theSystemIsUnderPeakLoadConditions() {
        simulatePeakLoad();
    }

    @When("system resources are monitored")
    public void systemResourcesAreMonitored() {
        monitorSystemResources();
    }

    @Then("system resources are utilized efficiently without bottlenecks")
    public void systemResourcesAreUtilizedEfficientlyWithoutBottlenecks() {
        verifyResourceUtilization();
    }

    @Given("multiple user sessions are active")
    public void multipleUserSessionsAreActive() {
        activateMultipleSessions();
    }

    @When("the response time for each session is measured")
    public void theResponseTimeForEachSessionIsMeasured() {
        measureResponseTime();
    }

    @Then("the response time remains within acceptable limits for all sessions")
    public void theResponseTimeRemainsWithinAcceptableLimitsForAllSessions() {
        verifyResponseTimeLimits();
    }

    @When("the UI responsiveness is checked for each session")
    public void theUIResponsivenessIsCheckedForEachSession() {
        checkUIResponsiveness();
    }

    @Then("the UI remains responsive and functional for all users")
    public void theUIRemainsResponsiveAndFunctionalForAllUsers() {
        verifyUIFunctionality();
    }

    @When("users attempt to access product details")
    public void usersAttemptToAccessProductDetails() {
        accessProductDetails();
    }

    @Then("product details are accessible without delay")
    public void productDetailsAreAccessibleWithoutDelay() {
        verifyProductDetailsAccessibility();
    }

    @Given("concurrent access is occurring")
    public void concurrentAccessIsOccurring() {
        simulateConcurrentAccess();
    }

    @When("system logs are observed")
    public void systemLogsAreObserved() {
        observeSystemLogs();
    }

    @Then("system logs do not contain critical errors related to concurrent access")
    public void systemLogsDoNotContainCriticalErrorsRelatedToConcurrentAccess() {
        verifySystemLogs();
    }

    @Given("the system has been under peak load")
    public void theSystemHasBeenUnderPeakLoad() {
        maintainPeakLoad();
    }

    @When("the load decreases")
    public void theLoadDecreases() {
        decreaseLoad();
    }

    @Then("the system recovers swiftly and continues to function optimally post-peak load")
    public void theSystemRecoversSwiftlyAndContinuesToFunctionOptimallyPostPeakLoad() {
        verifySystemRecovery();
    }

    @Given("varying network conditions")
    public void varyingNetworkConditions() {
        simulateNetworkConditions();
    }

    @When("system performance is tested")
    public void systemPerformanceIsTested() {
        testSystemPerformance();
    }

    @Then("the system performs consistently across varying network conditions")
    public void theSystemPerformsConsistentlyAcrossVaryingNetworkConditions() {
        verifyPerformanceConsistency();
    }

    @Given("the current user load")
    public void theCurrentUserLoad() {
        assessCurrentUserLoad();
    }

    @When("additional users are added")
    public void additionalUsersAreAdded() {
        addAdditionalUsers();
    }

    @Then("the system scales efficiently to handle additional user load")
    public void theSystemScalesEfficientlyToHandleAdditionalUserLoad() {
        verifySystemScalability();
    }

    @When("all user sessions are logged out")
    public void allUserSessionsAreLoggedOut() {
        logoutAllSessions();
    }

    @Then("all users are successfully logged out without errors")
    public void allUsersAreSuccessfullyLoggedOutWithoutErrors() {
        verifyLogoutSuccess();
    }

    @Given("different user accounts")
    public void differentUserAccounts() {
        setupDifferentUserAccounts();
    }

    @When("performance tests are repeated")
    public void performanceTestsAreRepeated() {
        repeatPerformanceTests();
    }

    @Then("system performance is consistent across different user accounts")
    public void systemPerformanceIsConsistentAcrossDifferentUserAccounts() {
        verifyPerformanceAcrossAccounts();
    }

    @Given("supported browsers")
    public void supportedBrowsers() {
        identifySupportedBrowsers();
    }

    @When("system performance is tested on each browser")
    public void systemPerformanceIsTestedOnEachBrowser() {
        testPerformanceOnBrowsers();
    }

    @Then("the system performs consistently on all supported browsers")
    public void theSystemPerformsConsistentlyOnAllSupportedBrowsers() {
        verifyBrowserPerformance();
    }

    @Given("mobile devices during peak load")
    public void mobileDevicesDuringPeakLoad() {
        setupMobileDevices();
    }

    @When("system performance is tested")
    public void systemPerformanceIsTestedOnMobile() {
        testPerformanceOnMobile();
    }

    @Then("the system performs optimally on mobile devices")
    public void theSystemPerformsOptimallyOnMobileDevices() {
        verifyMobilePerformance();
    }

    @Given("peak load conditions")
    public void peakLoadConditions() {
        simulatePeakLoadConditions();
    }

    @When("network latency is simulated")
    public void networkLatencyIsSimulated() {
        simulateNetworkLatency();
    }

    @Then("the system handles network latency gracefully without significant performance degradation")
    public void theSystemHandlesNetworkLatencyGracefullyWithoutSignificantPerformanceDegradation() {
        verifyLatencyHandling();
    }

    @Given("unexpected spikes in user load")
    public void unexpectedSpikesInUserLoad() {
        simulateUnexpectedSpikes();
    }

    @When("the system is tested")
    public void theSystemIsTested() {
        testSystemUnderSpikes();
    }

    @Then("the system manages unexpected spikes efficiently without crashing")
    public void theSystemManagesUnexpectedSpikesEfficientlyWithoutCrashing() {
        verifySpikeManagement();
    }
}

File: step_definitions/Verify system performance with maximum concurrent users accessing the catalog.js
import { Given, When, Then } from "cypress-cucumber-preprocessor/steps";
import SystemPerformancePage from "../pages/systemPerformancePage";

Given("the system is prepared for maximum concurrent user load", () => {
  SystemPerformancePage.prepareForMaxLoad();
});

When("concurrent user sessions are initiated to access the product catalog", () => {
  SystemPerformancePage.initiateConcurrentSessions();
});

Then("the system handles multiple user sessions without crashing", () => {
  SystemPerformancePage.verifySystemStability();
});

Then("the product catalog must be accessible", () => {
  SystemPerformancePage.verifyCatalogAccessibility();
});

Given("the system is under peak load conditions", () => {
  SystemPerformancePage.simulatePeakLoad();
});

When("system resources are monitored", () => {
  SystemPerformancePage.monitorResources();
});

Then("system resources are utilized efficiently without bottlenecks", () => {
  SystemPerformancePage.verifyResourceUtilization();
});

Given("multiple user sessions are active", () => {
  SystemPerformancePage.activateUserSessions();
});

When("the response time for each session is measured", () => {
  SystemPerformancePage.measureResponseTime();
});

Then("the response time remains within acceptable limits for all sessions", () => {
  SystemPerformancePage.verifyResponseTimeLimits();
});

When("the UI responsiveness is checked for each session", () => {
  SystemPerformancePage.checkUIResponsiveness();
});

Then("the UI remains responsive and functional for all users", () => {
  SystemPerformancePage.verifyUIResponsiveness();
});

When("users attempt to access product details", () => {
  SystemPerformancePage.accessProductDetails();
});

Then("product details are accessible without delay", () => {
  SystemPerformancePage.verifyProductDetailsAccessibility();
});

Given("concurrent access is occurring", () => {
  SystemPerformancePage.simulateConcurrentAccess();
});

When("system logs are observed", () => {
  SystemPerformancePage.observeSystemLogs();
});

Then("system logs do not contain critical errors related to concurrent access", () => {
  SystemPerformancePage.verifyLogErrors();
});

Given("the system has been under peak load", () => {
  SystemPerformancePage.simulatePeakLoad();
});

When("the load decreases", () => {
  SystemPerformancePage.decreaseLoad();
});

Then("the system recovers swiftly and continues to function optimally post-peak load", () => {
  SystemPerformancePage.verifySystemRecovery();
});

Given("varying network conditions", () => {
  SystemPerformancePage.simulateNetworkConditions();
});

When("system performance is tested", () => {
  SystemPerformancePage.testPerformance();
});

Then("the system performs consistently across varying network conditions", () => {
  SystemPerformancePage.verifyNetworkPerformance();
});

Given("the current user load", () => {
  SystemPerformancePage.observeCurrentLoad();
});

When("additional users are added", () => {
  SystemPerformancePage.addAdditionalUsers();
});

Then("the system scales efficiently to handle additional user load", () => {
  SystemPerformancePage.verifyScalability();
});

When("all user sessions are logged out", () => {
  SystemPerformancePage.logoutAllSessions();
});

Then("all users are successfully logged out without errors", () => {
  SystemPerformancePage.verifyLogoutSuccess();
});

Given("different user accounts", () => {
  SystemPerformancePage.selectDifferentAccounts();
});

When("performance tests are repeated", () => {
  SystemPerformancePage.repeatPerformanceTests();
});

Then("system performance is consistent across different user accounts", () => {
  SystemPerformancePage.verifyAccountPerformance();
});

Given("supported browsers", () => {
  SystemPerformancePage.selectSupportedBrowsers();
});

When("system performance is tested on each browser", () => {
  SystemPerformancePage.testOnBrowsers();
});

Then("the system performs consistently on all supported browsers", () => {
  SystemPerformancePage.verifyBrowserPerformance();
});

Given("mobile devices during peak load", () => {
  SystemPerformancePage.simulateMobilePeakLoad();
});

Then("the system performs optimally on mobile devices", () => {
  SystemPerformancePage.verifyMobilePerformance();
});

Given("peak load conditions", () => {
  SystemPerformancePage.simulatePeakLoad();
});

When("network latency is simulated", () => {
  SystemPerformancePage.simulateNetworkLatency();
});

Then("the system handles network latency gracefully without significant performance degradation", () => {
  SystemPerformancePage.verifyLatencyHandling();
});

Given("unexpected spikes in user load", () => {
  SystemPerformancePage.simulateUserSpikes();
});

When("the system is tested", () => {
  SystemPerformancePage.testSystemUnderSpikes();
});

Then("the system manages unexpected spikes efficiently without crashing", () => {
  SystemPerformancePage.verifySpikeManagement();
});